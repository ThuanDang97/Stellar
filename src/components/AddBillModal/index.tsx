import React from 'react'
import { useTheme } from 'styled-components'
import _ from 'lodash'

// Components
import AutoComplete from '@components/AutoComplete'
import Button from '@components/Button'
import Input from '@components/Input'
import Modal from '@components/Modal'
import { ViewStyled } from '@components/styled-components'
import { TitlePageStyled } from '@components/Title/TitleStyled'

// Props type
import { AddBillModalProps } from '@self-types/components/AddBillModal.props'

// Constants
import { LIST_BILLER } from '@constants/variables'

// Utils
import { findItemByValue } from '@utils/index'

const AddBillModal = ({
  isOpen,
  onClose,
  onClick,
  amount,
  handleChangeValue,
  billerItem,
  setBillerItem,
  onSubmit,
}: AddBillModalProps) => {
  const theme = useTheme()

  const billItem =
    findItemByValue({
      data: LIST_BILLER,
      value: billerItem,
      key: 'title',
    }) || {}

  const isDisabled = _.isEmpty(billItem) || !amount

  return (
    <Modal minWidth={theme.metrics.width.xl} isOpen={isOpen} onClose={onClose}>
      <ViewStyled
        display="flex"
        as="form"
        justifyContent="center"
        flexDirection="column"
        onSubmit={onSubmit}
        alignItems="initial"
        width={theme.metrics.width.full}
        pLeft={theme.metrics.dimensions.base}
        pRight={theme.metrics.dimensions.base}
      >
        <TitlePageStyled
          fontFamily="AdobeCleanExtraBold"
          color={theme.colors.black}
          fontSize={theme.typography.fontSize.xlg}
          fontWeight={theme.typography.fontSize.xlg}
          pBottom={theme.metrics.dimensions.base}
          pLeft={theme.metrics.dimensions.xmd}
          pRight={theme.metrics.dimensions.xmd}
        >
          Add Your Bill
        </TitlePageStyled>
        <AutoComplete billerItem={billerItem} setBillerItem={setBillerItem} />
        <Input
          type="text"
          value={amount}
          iconUrl="/icons/currency-dollar"
          onChange={handleChangeValue}
          name="amount"
          placeholder="Enter Amount"
          mTop={theme.metrics.dimensions.base}
          mBottom={theme.metrics.dimensions.sm}
          maxLength={12}
        />
        <Button
          {...(!isDisabled && { cursor: 'pointer' })}
          title="Confirm & Add Bill"
          type="submit"
          disabled={isDisabled}
          btnHeight={45}
          letterSpacing={theme.typography.letterSpacing.xl}
          onClick={onClick}
          mTop={theme.metrics.dimensions.xl}
          mBottom={theme.metrics.dimensions.base}
        />
      </ViewStyled>
    </Modal>
  )
}

export default React.memo(AddBillModal)
