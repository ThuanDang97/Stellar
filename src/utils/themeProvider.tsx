import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'
import { theme } from '@themes/index'

export const renderWithTheme = (component: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
}
