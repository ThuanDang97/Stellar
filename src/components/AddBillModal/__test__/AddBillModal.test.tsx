import React from 'react'
import { screen, cleanup } from '@testing-library/react'

// Utils
import { renderWithTheme } from '@utils/themeProvider'

// Components
import AddBillModal from '../index'

describe('AddBillModal', () => {
  const props = {
    onClose: jest.fn(),
    handleChangeValue: jest.fn(),
    amount: '',
    billerItem: '',
    setBillerItem: jest.fn(),
  }
  beforeEach(() => {
    renderWithTheme(<AddBillModal isOpen {...props} />)
  })

  afterEach(() => {
    cleanup()
  })

  it('should render AddBillModal', () => {
    const backdrop = screen.getByTestId('backdrop')

    expect(backdrop).toBeInTheDocument()
  })

  it('matches snapshot', () => {
    const { asFragment } = renderWithTheme(<AddBillModal isOpen {...props} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('should disable btn', () => {
    cleanup()
    renderWithTheme(<AddBillModal isOpen {...props} />)

    const button = screen.getByTestId('button')

    expect(button).toHaveAttribute('disabled')
  })

  it('should enable btn', () => {
    const newProps = { ...props, amount: '$1', billerItem: 'Netflix' }

    cleanup()
    renderWithTheme(<AddBillModal isOpen {...newProps} />)

    const button = screen.getByTestId('button')

    expect(button).not.toHaveAttribute('disabled')
  })
})
