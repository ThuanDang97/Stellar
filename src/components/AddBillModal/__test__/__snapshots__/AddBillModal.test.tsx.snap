// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddBillModal matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ViewStyled-sc-rqe5sa-0 ModalStyled__BackDropStyled-sc-d5vki3-1 fTsvHV cnMCGd"
    data-testid="backdrop"
  >
    <div
      class="ViewStyled-sc-rqe5sa-0 ViewStyled__BoxShadowStyled-sc-rqe5sa-2 ModalStyled-sc-d5vki3-0 eUUuqy hTNOqC hmXhmG"
    >
      <div
        class="ModalStyled__ImageWrapper-sc-d5vki3-2 eSAkbC"
        data-testid="close-icon"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
        >
          <img
            alt="close-icon"
            data-nimg="fill"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </div>
      <form
        class="ViewStyled-sc-rqe5sa-0 bbtxTf"
        display="flex"
        width="100%"
      >
        <span
          class="TextStyled-sc-9xia41-0 TitleStyled-sc-t58pjn-0 TitleStyled__TitlePageStyled-sc-t58pjn-1 jdHDfi epvbhg ilgUUT"
          color="#000"
          font-family="AdobeCleanExtraBold"
          font-size="36"
          font-weight="36"
        >
          Add Your Bill
        </span>
        <div
          class="ViewStyled-sc-rqe5sa-0 frRnNW"
          width="100%"
        >
          <div
            class="ViewStyled-sc-rqe5sa-0 ViewStyled__BorderWrapperStyled-sc-rqe5sa-4 kuBhDh eWayqq"
            data-testid="select"
            width="100%"
          >
            <div
              class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperInputStyled-sc-9gj86m-1 dUEIVo brcjID"
              color="#203F59"
              data-testid="inputIcon"
              display="block"
            >
              <div
                class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperIconStyled-sc-9gj86m-2 fTsvHV kHphsF"
              >
                <span
                  style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
                >
                  <img
                    alt="icon"
                    data-nimg="fill"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
              <input
                class="InputStyled-sc-1k09cqb-0 InputStyled__InputItemStyled-sc-9gj86m-0 hnxmhM NDCRX"
                data-testid="iconInput"
                placeholder="Select Biller"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperInputStyled-sc-9gj86m-1 eVnnax brcjID"
          color="#203F59"
          data-testid="inputIcon"
          display="block"
        >
          <div
            class="ViewStyled-sc-rqe5sa-0 InputStyled__WrapperIconStyled-sc-9gj86m-2 fTsvHV kHphsF"
          >
            <span
              style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
            >
              <img
                alt="icon"
                data-nimg="fill"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </div>
          <input
            class="InputStyled-sc-1k09cqb-0 InputStyled__InputItemStyled-sc-9gj86m-0 hnxmhM NDCRX"
            data-testid="iconInput"
            maxlength="12"
            name="amount"
            placeholder="Enter Amount"
            type="text"
            value=""
          />
        </div>
        <button
          class="ButtonStyled-sc-1ys9v52-0 kIqbUd"
          data-testid="button"
          disabled=""
          letter-spacing="0.69"
          type="submit"
        >
          Confirm & Add Bill
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
